### ISSUE TITLE
Cluster dies when stress testing

### ISSUE THREAD

**Issue by mattiasmak** [cockroachdb/cockroach#16080]
I've been using cockroachdb in two development projects. First project we had problems issue: #15026 Now we have problems again with another project/client.

For week everything runs smoothly. But when doing some stress tests the nodes started to go down and it's really hard to get them back up again. Since this is the second time in different project and teams we've had similar problems this is something that needs to be considered alarming. Perhaps it's us who don't have the technical skills to setup the cluster right? But it seems very straight forward. 

The stress tests was on the GraphQL endpoint which in turn did a few sql-queries. Not many. The test failed after just about 10 simultaneous requests. We've done a few other during development. The only new thing was this query:

`SELECT id FROM tgw.user WHERE lower(name) LIKE '%${search}%' LIMIT ${limit};`

I'm happy to share whatever you need to figure this out. I attach the logs files. This problem happened at 2015-05-22

[create_db.txt](https://github.com/cockroachdb/cockroach/files/1022114/create_db.txt)

[logs-003.tar.gz](https://github.com/cockroachdb/cockroach/files/1022104/logs-003.tar.gz)
[logs-002.tar.gz](https://github.com/cockroachdb/cockroach/files/1022105/logs-002.tar.gz)
[logs-001.tar.gz](https://github.com/cockroachdb/cockroach/files/1022106/logs-001.tar.gz)

**Comment by petermattis** [comment]
@tschottdorf Can you take a look at this?

**Comment by tschottdorf** [comment]
Looking.

**Comment by tschottdorf** [comment]
Hi @mattiasmak, from what it looks like to me the cluster is unhappy because it lost a majority of the system ranges (and in fact, likely many other ranges). From the logs, I can gather the following timeline:

5/19: node1 is cleanly shutdown
5/22 12:56pm: node1 is restarted
5/22 12:45pm: node3 crashes with out-of-memory but restarts at 12:56
5/22 13:02pm: node2 stops (or crashes; hard to tell)


What happens behind the scenes is that on 5/19, after the remaining nodes decide that node1 is dead, they remove node1 from the replication. You are now running with all ranges under-replicated (if there were another node in the system, that node would receive a replica and the cluster would be happy, but there is no fourth node). When node1 comes back, it doesn't have enough time to receive these replicas again. Very shortly after, node2 goes away, and so now a lot of ranges have only one surviving replica: node3. In particular, you lose the vital system ranges, so the cluster is in a problematic state. Note that I didn't even use the fact that node3 is offline between 12:45 and 12:56. During that period, the cluster is also very likely very broken as you also lose majorities, and since that period is larger than 5 minutes, if the cluster worked before, it may have gotten itself into more trouble by taking replicas from node 3 to node 1.

As for the reason for the crashes during stress testing, it seems to me that you're running pretty tight on memory. Found this in node3's log (perhaps this is what happened during the stress test):

```
I170522 12:44:53.494980 68 server/status/runtime.go:225  [n3] runtime stats: 827 MiB RSS, 914 goroutines, 294 MiB/60 MiB
/415 MiB GO alloc/idle/total, 266 MiB/334 MiB CGO alloc/total, 267.39cgo/sec, 0.10/0.29 %(u/s)time, 0.00 %gc (0x)
...
[n3,s3,r42/3:/System/tsd/cr.node.sys.cgoca…] new range leas
e repl=(n3,s3):3 start=1495210715.812461553,0 epo=3 pro=1495457102.325673999,0 following repl=(n3,s3):3 start=1495210715
.812461553,0 epo=2 pro=1495210715.812470584,0
fatal error: runtime: out of memory

runtime stack:
runtime.throw(0x1b18830, 0x16)
        /usr/local/go/src/runtime/panic.go:596 +0x95
runtime.sysMap(0xc438d00000, 0x100000, 0x500, 0x2c43f78)
        /usr/local/go/src/runtime/mem_linux.go:216 +0x1d0
```

Note the line

```
I170522 12:44:53.494980 68 server/status/runtime.go:225  [n3] runtime stats: 827 MiB RSS, 914 goroutines, 294 MiB/60 MiB
/415 MiB GO alloc/idle/total, 266 MiB/334 MiB CGO alloc/total, 267.39cgo/sec, 0.10/0.29 %(u/s)time, 0.00 %gc (0x)
```

which indicates that before you start the test, the node looks pretty idle. The crash happens ten seconds later. Unless your workload somehow tickles something really bad in CockroachDB, my suggestion would be to share with us how much memory these nodes are running on. If that looks ok, we could try out the actual workload.

Unrelated, but you seem to be running v1.0-rc2. I wasn't able to find an obvious commit between 1.0-rc2 and 1.0 that would make a difference here, but I'd nevertheless suggest updating to v1.0 (or the soon-to-be-released v1.0.1) whenever you get a chance.


Everything below are analysis notes, which I don't think are very illuminating except for @a-robinson who may want to take a look for anything that looks odd to him.

--


It's pretty clear from looking at the logs that gossip is unhealthy for *most* of the time covered in the logs from 5/22 on. Node 1 is restarted repeatedly around 2017-05-22T12_47 (presumably in an attempt to resolve the situation). Interestingly, Node 1 was actually cleanly stopped between 170519 and the restarts, so the cluster was running with a dead node. Gossip on the remaining two nodes seemed happy, though, until node 1 came back.

The problems start when node1 comes back online (and node3 is restarted), at around 2017-05-22T12_55.

on startup, node 3 goes into a loop of:

```
I170522 12:57:36.830974 2783 gossip/client.go:131  [n3] started gossip client to 10.135.67.192:26257
I170522 12:57:36.834020 2783 gossip/client.go:136  [n3] closing client to node 1 (10.135.67.192:26257): stopping outgoing client to node 1 (10.135.67.192:26257); already have incoming
I170522 12:57:37.831414 2813 gossip/client.go:131  [n3] started gossip client to 10.135.70.106:26257
I170522 12:57:38.661065 2813 gossip/client.go:136  [n3] closing client to node 2 (10.135.70.106:26257): stopping outgoing client to node 2 (10.135.70.106:26257); already have incoming
```

and its gossip status shows no active connections:

```
I170522 12:57:34.979339 71 gossip/gossip.go:448  [n3] gossip status (stalled, 3 nodes)
```

but after around a minute of that, it miraculously connects (by virtue of having received inbound connections from nodes 2 and 1).

```
I170522 12:58:07.644774 3504 gossip/client.go:131  [n3] started gossip client to 10.135.67.192:26257
I170522 12:58:07.646202 3504 gossip/client.go:136  [n3] closing client to node 1 (10.135.67.192:26257): stopping outgoing client to node 1 (10.135.67.192:26257); already have incoming
I170522 12:58:07.646282 27 gossip/gossip.go:1203  [n3] node has connected to cluster via gossip
...
I170522 12:58:34.981763 71 gossip/gossip.go:448  [n3] gossip status (ok, 3 nodes)
gossip client (0/3 cur/max conns)
gossip server (2/3 cur/max conns, infos 2538/179 sent/received, bytes 1821632B/103024B sent/received)
  2: 10.135.70.106:26257 (2m0s)
  1: 10.135.67.192:26257 (2m0s)
```

At 13:02, node2 goes down (it does not look like a clean shutdown; not sure what occurred there; the node does seem pretty busy with snapshots and the like, showing some slow RocksDB activity, but it's runtime stats look ok:

```
I170522 13:02:02.545098 55 server/status/runtime.go:225  [n2] runtime stats: 821 MiB RSS, 953 goroutines, 272 MiB/80 MiB
/389 MiB GO alloc/idle/total, 262 MiB/332 MiB CGO alloc/total, 6181.39cgo/sec, 0.15/0.18 %(u/s)time, 0.00 %gc (0x)
```

Either way, node3 goes back to an infinite loop, thrashing only in n1's direction.

```
gossip client (0/3 cur/max conns)
gossip server (1/3 cur/max conns, infos 3128/364 sent/received, bytes 1903788B/136609B sent/received)
  1: 10.135.67.192:26257 (6m0s)
I170522 13:04:01.919213 9451 gossip/client.go:131  [n3] started gossip client to 10.135.67.192:26257
I170522 13:04:01.924770 9451 gossip/client.go:136  [n3] closing client to node 1 (10.135.67.192:26257): stopping outgoing client to node 1 (10.135.67.192:26257); already have incoming
I170522 13:04:03.919961 9474 gossip/client.go:131  [n3] started gossip client to 10.135.67.192:26257
I170522 13:04:03.921995 9474 gossip/client.go:136  [n3] closing client to node 1 (10.135.67.192:26257): stopping outgoing client to node 1 (10.135.67.192:26257); already have incoming
```

node1 in the meantime is busy removing a lot of its replicas that it had shed after restarting after three days of downtime - the rest of the cluster had rebalanced away all of its replicas at that point:

```
I170522 13:02:02.974060 55 storage/replica.go:673  [replicaGC,n1,s1,r50/1:/System/tsd/cr.node.{ro…-sq…}] removed 5902 (5891+11) keys in 12ms [clear=12ms commit=0ms]
I170522 13:02:02.974060 55 storage/replica.go:673  [replicaGC,n1,s1,r50/1:/System/tsd/cr.node.{ro…-sq…}] removed 5902 (5891+11) keys in 12ms [clear=12ms commit=0ms]
I170522 13:02:03.010232 4316 storage/store.go:3149  [n1,s1,r94/1:/System/tsd/cr.node.sql.mem.…] added to replica GC queue (peer suggestion)
I170522 13:02:03.010232 4316 storage/store.go:3149  [n1,s1,r94/1:/System/tsd/cr.node.sql.mem.…] added to replica GC queue (peer suggestion)
I170522 13:02:03.013201 4316 storage/store.go:3149  [n1,s1,r59/1:/System/tsd/cr.node.sql.mem.c…] added to replica GC queue (peer suggestion)
I170522 13:02:03.013201 4316 storage/store.go:3149  [n1,s1,r59/1:/System/tsd/cr.node.sql.mem.c…] added to replica GC queue (peer suggestion)
```

As far as gossip is concerned, it's doing OK until 13:02 (when node2 goes away):

```
I170522 13:02:13.356568 62 gossip/gossip.go:448  [n1] gossip status (ok, 3 nodes)
gossip client (2/3 cur/max conns)
  2: 10.135.70.106:26257 (6m59s: infos 387/159 sent/received, bytes 78739B/35001B sent/received)
  3: 10.135.70.165:26257 (5m39s: infos 115/527 sent/received, bytes 51270B/66973B sent/received)
gossip server (0/3 cur/max conns, infos 539/3372 sent/received, bytes 161045B/2255753B sent/received)
W170522 13:04:03.728812 27 gossip/gossip.go:1194  [n1] first range unavailable; trying remaining resolvers
W170522 13:04:03.729796 5656 gossip/client.go:125  [n1] failed to start gossip client to 10.135.70.106:26257: rpc error: code = Unavailable desc = grpc: the connection is unavailable
I170522 13:04:13.366553 63 server/status/runtime.go:225  [n1] runtime stats: 136 MiB RSS, 151 goroutines, 42 MiB/22 MiB/73 MiB GO alloc/idle/total, 20 MiB/23 MiB CGO alloc/total, 65.05cgo/sec, 0.02/0.01 %(u/s)time, 0.00 %gc (0x)
I170522 13:04:13.368069 62 gossip/gossip.go:448  [n1] gossip status (stalled, 3 nodes)
gossip client (2/3 cur/max conns)
  3: 10.135.70.165:26257 (7m39s: infos 117/533 sent/received, bytes 51682B/67895B sent/received)
  0: 10.135.70.106:26257 (3s: infos 0/0 sent/received, bytes 0B/0B sent/received)
```

**Comment by tschottdorf** [comment]
As an aside here, it seems somewhat problematic that we would downreplicate in a three node cluster when one node is unavailable as it can increase recovery times in case the same node reappears later. Had we not done that, I think this case wouldn't have been so terrible (though in this scenario you certainly can't expect to survive). I need to double-check that that's what we're actually doing, but the logs strongly suggest so.

**Comment by mattiasmak** [comment]
The nodes run on digital ocean:  1 GB Memory / 30 GB Disk / Ubuntu 16.04.2 x64 

Please let me know if there's something I can do? Do you want access to the nodes? Is this something we made wrong? I did everything by the book, so I hope not. 

I just want to make sure that this will not happen once we're in production. :/

**Comment by mattiasmak** [comment]
There's a lot of information to take in. But if there's something we can do in order to avoid this I come up with:

1. Make sure there's supervision on the cluster alarming of one node goes down.
2. More memory than 1 GB on the nodes
3. Perhaps at least 5 nodes

Is the above point good ideas? I guess there's something problematic going on with cockroach as well. I think what's bothering me a bit is that so hard to recover from the current state. Since it's only development I guess I will kill the nodes and start a new cluster. But as I said. What if we were in production?

**Comment by tschottdorf** [comment]
Hi @mattiasmak, 1. and 2. sound good. I'm not so sure about 3. - going to 5 nodes doesn't increase your availability much (unless you also run with five-fold replication as opposed to the default of three).

It's possible that CockroachDB itself is causing some trouble. Unfortunately, I can't really infer from the logs if that's the case because the ones you provided for node2 stop too early (i.e. I don't ever see you restarting that node), and what I see indicate that you've practically lost two out of three nodes by having node1 offline for a long time and then just after it comes back taking node2 offline (CockroachDB can do a little better there, but no consistent system will be able to give you a lot of guarantees if you do that).

That said, I assume you *did* in fact restart all three nodes and it still didn't work. I'm happy to look into that for you. As far as I can tell, it should be the case that once all three processes are up and running things go back to mostly normal, assuming the shortage of ram isn't too severe.

If you were able to arrange access to that cluster, that would be the easiest option - perhaps send me a message on gitter.im and we'll work through it from there!

**Comment by bdarnell** [comment]
> it seems somewhat problematic that we would downreplicate in a three node cluster when one node is unavailable as it can increase recovery times in case the same node reappears later

Yeah, downreplicating from unavailable nodes only helps if it's a precursor to upreplicating elsewhere. We should avoid removing dead replicas unless there are suitable live targets. I think running 5 nodes instead of 3 (even if the replication factor remains at 3) would improve availability for this reason: we'd have somewhere to put new replicas after one node goes down and wouldn't stay in an underreplicated state. 

It's been a while since we've done much testing of a 3-node cluster; maybe we should add a test configuration for 3 nodes with chaos. 

> That said, I assume you did in fact restart all three nodes and it still didn't work. I'm happy to look into that for you. As far as I can tell, it should be the case that once all three processes are up and running things go back to mostly normal, assuming the shortage of ram isn't too severe.

Yes, if the cluster not able to recover after all three nodes have been restarted (with their disks intact), then that's very concerning.

**Comment by tschottdorf** [comment]
> I think running 5 nodes instead of 3 (even if the replication factor remains at 3) would improve availability for this reason: we'd have somewhere to put new replicas after one node goes down and wouldn't stay in an underreplicated state.

Ah yes, you're right. No reason to go up to 5 though, right? Could simply do 4.

I'll file issues for not downreplicating when there's no upreplication target && running a 3node cluster chaos test.

**Comment by bdarnell** [comment]
Yeah, 4 would be enough if you're sticking with 3-way replication.

**Comment by tschottdorf** [comment]
@mattiasmak as a follow-up, https://github.com/cockroachdb/cockroach/issues/16097 has been addressed, which means that the window of vulnerability in these three-node clusters has been somewhat reduced (though losing a quorum is still not acceptable). It will be available in 1.0.2.

I'll close this issue, but feel free to ping back with any updates.
